exchanges/okx/okx.go

```diff
@@ -0,0 +1,4496 @@
+		if err != nil {
+			return nil, err
+		}
+		takerVolume := TakerVolume{
+			Timestamp:  time.UnixMilli(int64(timestamp)),
+			SellVolume: sellVolume,
+			BuyVolume:  buyVolume,
+		}
+		takerVolumes = append(takerVolumes, takerVolume)
+	}
+	return takerVolumes, nil
+}
+
+// GetMarginLendingRatio retrieves the ratio of cumulative amount between currency margin quote currency and base currency.
+func (ok *Okx) GetMarginLendingRatio(ctx context.Context, currency string, begin, end time.Time, period kline.Interval) ([]MarginLendRatioItem, error) {
+	params := url.Values{}
+	if currency != "" {
+		params.Set("ccy", currency)
+	}
+	if !begin.IsZero() {
+		params.Set("begin", strconv.FormatInt(begin.UnixMilli(), 10))
+	}
+	if !end.IsZero() {
+		params.Set("end", strconv.FormatInt(begin.UnixMilli(), 10))
+	}
+	interval := ok.GetIntervalEnum(period)
+	if interval != "" {
+		params.Set("period", interval)
+	}
+	var response [][2]string
+	err := ok.SendHTTPRequest(ctx, exchange.RestSpot, getMarginLendingRatioEPL, http.MethodGet, common.EncodeURLValues(tradingMarginLoanRatio, params), nil, &response, false)
+	if err != nil {
+		return nil, err
+	}
+	lendingRatios := []MarginLendRatioItem{}
+	for x := range response {
+		if len(response[x]) != 2 {
+			return nil, errMalformedData
+		}
+		timestamp, err := strconv.Atoi(response[x][0])
+		if err != nil {
+			return nil, err
+		}
+		ratio, err := strconv.ParseFloat(response[x][0], 64)
+		if err != nil || ratio <= 0 {
+			return nil, err
+		}
+		lendRatio := MarginLendRatioItem{
+			Timestamp:       time.UnixMilli(int64(timestamp)),
+			MarginLendRatio: ratio,
+		}
+		lendingRatios = append(lendingRatios, lendRatio)
+	}
+	return lendingRatios, nil
+}
+
+// GetLongShortRatio retrieves the ratio of users with net long vs net short positions for futures and perpetual swaps.
+func (ok *Okx) GetLongShortRatio(ctx context.Context, currency string, begin, end time.Time, period kline.Interval) ([]LongShortRatio, error) {
+	params := url.Values{}
+	if currency != "" {
+		params.Set("ccy", currency)
+	}
+	if !begin.IsZero() {
+		params.Set("begin", strconv.FormatInt(begin.UnixMilli(), 10))
+	}
+	if !end.IsZero() {
+		params.Set("end", strconv.FormatInt(begin.UnixMilli(), 10))
+	}
+	interval := ok.GetIntervalEnum(period)
+	if interval != "" {
+		params.Set("period", interval)
+	}
+	var response [][2]string
+	err := ok.SendHTTPRequest(ctx, exchange.RestSpot, getLongShortRatioEPL, http.MethodGet, common.EncodeURLValues(longShortAccountRatio, params), nil, &response, false)
+	if err != nil {
+		return nil, err
+	}
+	ratios := []LongShortRatio{}
+	for x := range response {
+		if len(response[x]) != 2 {
+			return nil, fmt.Errorf("%v, expecting length 2 but found %d", errMalformedData, len(response[x]))
+		}
+		timestamp, err := strconv.Atoi(response[x][0])
+		if err != nil || timestamp <= 0 {
+			return nil, err
+		}
+		ratio, err := strconv.ParseFloat(response[x][0], 64)
```

Comment:
this is also parsing the timestamp again. Also please remove the `ratio <= 0` as this will return nil with no data.